include/swift/IDE/SourceEntityWalker.h

```diff
@@ -82,11 +82,26 @@ class SourceEntityWalker {
 
   /// This method is called when first visiting a decl, before walking into its
   /// children.  If it returns false, the subtree is skipped.
-  virtual bool walkToDeclPre(Decl *D, CharSourceRange Range) { return true; }
+  virtual bool walkToDeclPre(Decl *D, CharSourceRange Range) {
+    beginBalancedDeclVisit(D);
```

Comment:
I added them in the case that a subclass doesn't override `walkToDeclPre/Post`, but does override `begin/endBalancedDeclVisit`. I added them for correctness really, currently they don't serve any practical benefit. I'm happy to remove them if you'd like.

Reply:
Won't this cause `beginBalancedDeclVisit` to get called twice if you don't override `walkToDeclPre` since SemaAnnotator will call it directly, then call `walkToDeclPre` which will come here and call it again?

Reply:
Ah yes, that's a good point. I'll remove these calls, and update the documentation for `walkToDeclPre/Post` to mention that it's the job of the subclass to implement calls to `begin/endBalancedDeclVisit`.

Reply:
No, you're quite right. It's me that was confused.